db/upgrade.php

```diff
@@ -109,5 +111,69 @@ function xmldb_tool_crawler_upgrade($oldversion) {
         upgrade_plugin_savepoint(true, 2020012300, 'tool', 'crawler');
     }
 
+    if ($oldversion < 2020030501) {
+        // Add persistent API fields to the 3 DB tables.
+        $tablenames = ['tool_crawler_url', 'tool_crawler_edge', 'tool_crawler_history'];
+        foreach ($tablenames as $tablename) {
+            $table = new xmldb_table($tablename);
+
+            if ($tablename === 'tool_crawler_url') {
+                // Use existing createdate field as timecreated instead of creating a new timecreated field.
+                $field = new xmldb_field('createdate', XMLDB_TYPE_INTEGER, '10', null, true, false);
+                if ($dbman->field_exists($table, $field)) {
+                    $dbman->rename_field($table, $field, 'timecreated');
+                }
+                // Add new urlhash field to the url table and index it.
+                $field = new xmldb_field('urlhash', XMLDB_TYPE_CHAR, '255', null, true, false);
+                $index = new xmldb_index('urlhash', XMLDB_INDEX_NOTUNIQUE, array('urlhash'));
+
+                if (!$dbman->field_exists($table, $field)) {
+                    $dbman->add_field($table, $field);
+                }
+                // Before indexing, populate the urlhash field with hashes of the currently existing urls.
+                core_php_time_limit::raise();
+                $urlrecords = $DB->get_recordset('tool_crawler_url');
```

Comment:
this will be crazy slow to churn through all the urls. Can this be done in sql?

SELECT encode(digest('<root/>'::bytea, 'sha1'), u.url)